qi-test/tests/flow.rkt

```diff
@@ -76,7 +76,10 @@
      (check-true ((☯ (and (> 5) (< 10))) 6))
      (check-false ((☯ (and (> 5) (< 10))) 4))
      (check-false ((☯ (and (> 5) (< 10))) 14))
-     (check-false ((☯ (and _ positive?)) #f) "_ in and"))
+     ;; TODO: this doesn't short-circuit anymore
+     ;; review shortcircuiting on all boolean forms
+     ;; (check-false ((☯ (and _ positive?)) #f) "_ in and")
```

Comment:
Yes, I was planning to restore it via a "restorative" compiler optimization. That is, there could be an optimization rule that matches this specific form:
```
(~> (-< onex ...) AND)
```
which is the current expansion for `and`, and rewrites it to
```
(conjoin (qi0->racket onex) ...)
```
... the original expansion for `and`.

Btw I haven't really written a compiler before (unless you count [C3PO](https://docs.racket-lang.org/social-contract/index.html#%28part._c3po%29)), so I'm making this stuff up as I go 😄

Reply:
Hm, are they really equivalent in the presence of side-effects?

Assuming `(~> (-< e ...) AND)` is the source form, it always executes all the side-effects of each `e`. The "optimization" would possibly prevent some side-effects _because_ of the short-circuiting. Concrete example:

```
> ((flow (~> (-< #f (esc (lambda _ (println "hello") #t))) AND)))
"hello"
#f
> ((conjoin (const #f) (lambda _ (println "hello") #t)))
#f
```

IOW the second transformation is valid
- when the original form was produced by the first (since short-circuiting is expected), or
- when all the sub-flows are side-effect free (since there is no observable difference, though the extra memory and computation time may be "observable").

Reply:
Ok, so this is going to be a lot harder than I thought 😅 

That all makes perfect sense. Could we do the transformation but in a more nuanced way, where we detect the presence of side effects and avoid making the optimization in those cases (e.g. maybe if either the `effect` form or `esc` is present, then don't transform it)? Seems like it would be great if we can treat `AND` and `OR` as core language constructs that all boolean-related forms reduce to.

Reply:
Waving my hands, detecting side-effects sounds like a Halting problem territory (I have nothing to back this up but an intuitive niggle, though).

Notably, `effect` and `esc` are not strictly necessary for effects:
```racket
(define (pure . _)
  (println "hello")
  #t)
((flow (~> (-< #f pure) AND)))
```

I haven't reasoned it all the way through yet, but I'm not sure I see how to build short-circuiting forms from non-short-circuiting forms (I consider `if` to be short-circuiting here, since it fully avoids evaluating one branch every time, and I think it's straightforward to build short-circuiting binary [n-ary] conjunction and disjunction with [nested] `if`). OTOH, it seems easy to build non-short-circuiting forms with short-circuiting forms (do both [all] evaluations, then apply the short-circuit operator to the results).

In that sense, I think the short-circuit operators need to be core, and the non-short-circuit operators can be implemented in terms of them. 

Not sure what optimizations might then be present (maybe in simple cases with only literals and Qi we can be sure about side-effects? but are there useful computations without invoking Racket procedures, which we cannot guarantee side-effect-free without knowing their origin?). Optimizing is essentially theorem proving 😅

Reply:
Yeah, good call. I restored short-circuiting `and` and `or` as core forms. My first attempts at writing `AND` in terms of `if` didn't bear fruit, e.g. I think this generates infinite code:
```
[(~or* (~datum AND) (~datum &))
 #'(qi0->racket (if live?
                    (if (~> (block 1) live?)
                        (if 1>
                            (group 1 ⏚ AND)
                            #f)
                        1>)
                    #t))]
```

Can revisit it later unless an obvious implementation strikes you.

Re: side effects, yeah, I'm not sure. If it came to it, at a high level, I wouldn't be opposed to weakening the guarantees provided by the language in some cases, e.g. cases where useful optimizations are possible _and_ the behavior is considered unidiomatic anyway. In such cases, we could advise that users use side effects explicitly via `effect` or `esc`. This is just abstract though, I don't have anything specific in mind. But also, weakening guarantees in specific cases are precisely how Qi could potentially even exceed Racket performance, so investing in idiomatic cases while supporting unidiomatic cases via `esc` and `effect` could be worth it. 🤷

Reply:
Well, this seems almost silly, but:
```
> (define-flow AND* (loop #t _ (and 1> 2>)))
> ((flow (~> (-< #f (esc (lambda _ (println "hi") #t))) AND)))
"hi"
#f
> (define-flow AND** (>> (and 1> 2>) #t))
> ;; print 1 and 2 to confirm execution and show order is left-to-right
> ((flow (~> (-< (esc (lambda _ (println 1) #f)) (esc (lambda _ (println 2) #t))) AND**)))
1
2
#f
```
I think `OR` would then be `(>> (or 1> 2>) #f)`.

---

RE: side-effects, I think you're right that restrictions lead to optimizations (in general, I think, the smaller the language, the more you know and therefore the more you can prove and therefore use for optimization; creativity under constraint à la OULIPO rings similar bells though has nothing to do with optimizing compilers). So I'm OK with saying effects must come under `effect`/`esc`, but it would sure be nice to remove or rule-out foot-guns with hidden effects like the `pure` example from above. Anyway, since that would be a breaking API change, that could be something we defer to later?

Reply:
Sure why not! Made the change. OULIPO is a fun analogy for compiler optimizations 😆